shared/contracts/BaseTemplate.sol

```diff
@@ -0,0 +1,191 @@
+import "@aragon/apps-finance/contracts/Finance.sol";
+import "@aragon/apps-token-manager/contracts/TokenManager.sol";
+import "@aragon/apps-shared-minime/contracts/MiniMeToken.sol";
+
+import "@aragon/os/contracts/acl/ACL.sol";
+import "@aragon/os/contracts/apm/Repo.sol";
+import "@aragon/os/contracts/apm/APMNamehash.sol";
+import "@aragon/os/contracts/kernel/Kernel.sol";
+import "@aragon/os/contracts/lib/ens/ENS.sol";
+import "@aragon/os/contracts/lib/ens/PublicResolver.sol";
+import "@aragon/os/contracts/factory/DAOFactory.sol";
+import "@aragon/os/contracts/common/IsContract.sol";
+import "@aragon/id/contracts/IFIFSResolvingRegistrar.sol";
+
+
+contract BaseTemplate is APMNamehash, IsContract {
+    /* Hardcoded constant to save gas
+    * bytes32 constant internal AGENT_APP_ID = apmNamehash("agent");                  // agent.aragonpm.eth
+    * bytes32 constant internal VAULT_APP_ID = apmNamehash("vault");                  // vault.aragonpm.eth
+    * bytes32 constant internal VOTING_APP_ID = apmNamehash("voting");                // voting.aragonpm.eth
+    * bytes32 constant internal FINANCE_APP_ID = apmNamehash("finance");              // finance.aragonpm.eth
+    * bytes32 constant internal TOKEN_MANAGER_APP_ID = apmNamehash("token-manager");  // token-manager.aragonpm.eth
+    */
+
+    bytes32 constant internal AGENT_APP_ID = 0x9ac98dc5f995bf0211ed589ef022719d1487e5cb2bab505676f0d084c07cf89a;
+    bytes32 constant internal VAULT_APP_ID = 0x7e852e0fcfce6551c13800f1e7476f982525c2b5277ba14b24339c68416336d1;
+    bytes32 constant internal VOTING_APP_ID = 0x9fa3927f639745e587912d4b0fea7ef9013bf93fb907d29faeab57417ba6e1d4;
+    bytes32 constant internal FINANCE_APP_ID = 0xbf8491150dafc5dcaee5b861414dca922de09ccffa344964ae167212e8c673ae;
+    bytes32 constant internal TOKEN_MANAGER_APP_ID = 0x6b20a3010614eeebf2138ccec99f028a61c811b3b1a3343b6ff635985c75c91f;
+
+    ENS public ens;
+    DAOFactory public daoFactory;
+    MiniMeTokenFactory public miniMeFactory;
+    IFIFSResolvingRegistrar public aragonID;
+
+    event DeployDao(address dao);
+    event DeployToken(address token);
+    event InstalledApp(address appProxy, bytes32 appId);
+
+    constructor(DAOFactory _daoFactory, ENS _ens, MiniMeTokenFactory _miniMeFactory, IFIFSResolvingRegistrar _aragonID) public {
+        require(isContract(address(_ens)), "TEMPLATE_ENS_NOT_CONTRACT");
+        require(isContract(address(_daoFactory)), "TEMPLATE_DAO_FACTORY_NOT_CONTRACT");
+
+        ens = _ens;
+        aragonID = _aragonID;
+        daoFactory = _daoFactory;
+        miniMeFactory = _miniMeFactory;
+    }
+
+    function _createDAO() internal returns (Kernel dao, ACL acl) {
+        dao = daoFactory.newDAO(this);
+        emit DeployDao(address(dao));
+        acl = ACL(dao.acl());
+        _createPermission(acl, dao, dao.APP_MANAGER_ROLE());
+    }
+
+    /* ACL */
+
+    function _createPermission(ACL acl, address app, bytes32 permission) internal {
```

Comment:
Perhaps we can add a `ToFactory` or `FromFactory` to these function names, to make it clear who the assignee is.